### PR TITLE
Fixed auto-loader to skip non-WP-Auth0 classes

### LINKED ISSUE #461: Autoloader performance issue
It appears that the autoloader slows down the wordpress site significantly.

Running performance monitor I was able to see that for every class that was being searched for, it would search in a bunch of different directories - even for files that were not part of the auth0 plugin. The more plugins I had installed, the worse the problem would get.

image

I believe the issue exists here:

https://github.com/auth0/wp-auth0/blob/9713e6a7cdf4c3446632be5338426dfba102bf3a/WP_Auth0.php#L412

And this is the fix that I did that seems to help things although I don't know if it is a permanent solution:

```
foreach ( $paths as $p ) {
    foreach ( $exts as $ext ) {
        if (strpos($p.$class.$ext, "WP_Auth0") !== FALSE && file_exists( $p.$class.$ext ) ) {
            require_once $p.$class.$ext;
            return true;
        }
    }
}
```

Thanks.

### PR DESCRIPTION
**Pull request by joshcanhelp**
Fixes #461

### CODE REVIEW

File: WP_Auth0.php [hunk 1]
```diff
@@ -436,27 +436,50 @@ public static function uninstall() {
...
+			case 'BeforeValidException':
+			case 'ExpiredException':
+			case 'SignatureInvalidException':
+				require_once $source_dir . 'php-jwt/Exceptions/'  . $class . '.php';
+				return true;
+		}
+
+		// Anything that's not part of the above and not name-spaced can be skipped.
+		if ( 0 !== strpos( $class, 'WP_Auth0' ) ) {
+			return false;
```
Comment (lbalmaceda) [inline WP_Auth0.php]: should you log something on this line?

File: WP_Auth0.php [hunk 2]
```diff
@@ -436,27 +436,50 @@ public static function uninstall() {
...
+		}
+
+		$paths = array(
+			$source_dir,
+			$source_dir . 'admin/',
+			$source_dir . 'exceptions/',
+			$source_dir . 'wizard/',
+			$source_dir . 'initial-setup/',
```
Comment (lbalmaceda) [inline WP_Auth0.php]: I guess this last comma will be ignored.

File: WP_Auth0.php [hunk 3]
```diff
@@ -436,27 +436,50 @@ public static function uninstall() {
...
+		);
+
+		foreach ( $paths as $path ) {
```
Comment (lbalmaceda) [inline WP_Auth0.php]: I don't know how often this method would be called, but I suppose several times per execution. Is there any way of iterating through all the files in a more efficient way? Maybe accepting an array of "classes to require" so you can iterate once and require them all.

File: WP_Auth0.php [hunk 4]
```diff
@@ -436,27 +436,50 @@ public static function uninstall() {
...
+			if ( file_exists( $path . $class . '.php' ) ) {
+				require_once $path . $class . '.php';
+				return true;
 			}
```
Comment (lbalmaceda) [inline WP_Auth0.php]: no need to `return false`?

File: WP_Auth0.php (hunk 1 above)
Comment (joshcanhelp) [reply WP_Auth0.php]: Definitely not. This is called for all class instantiations. It's not an error if an non-WP-Auth0 class is being called. We want to escape quickly.

File: WP_Auth0.php (hunk 2 above)
Comment (joshcanhelp) [reply WP_Auth0.php]: Yes. It's a PHP standard to add a comma on the last line of an array

File: WP_Auth0.php (hunk 3 above)
Comment (joshcanhelp) [reply WP_Auth0.php]: > several times per execution

Correct

> Maybe accepting an array of "classes to require"

That would defeat the purpose of the autoloader. This is here to "magically" include the necessary files for whatever class you might call. There is a more efficient way to do this using [Namespaces](https://secure.php.net/manual/en/language.namespaces.php) but that would require a lot more work (I have it on my list)

File: WP_Auth0.php (hunk 4 above)
Comment (joshcanhelp) [reply WP_Auth0.php]: Nope!

**Pull request merged**